### DISCUSSION TITLE
Sample data

### DISCUSSION THREAD

**Discussion by roaldarbol** [mooch443/trex#228] [General]
Hi @mooch443! Hope you're well!
I'm writing an R package for analysing movement data from tracking such as TRex ([animovement](https://roald-arboel.com/animovement)). Has the data format (headers/columns) changed from v1 to v2? If that's the case, do you happen to have some sample data I could use to write a TRex v2 reader too?
Cheers, Mikkel.

**Comment by mooch443** [comment]
> Hi @mooch443! Hope you're well!
> 
> I'm writing an R package for analysing movement data from tracking such as TRex ([animovement](https://roald-arboel.com/animovement)). Has the data format (headers/columns) changed from v1 to v2? If that's the case, do you happen to have some sample data I could use to write a TRex v2 reader too?
> 
> Cheers, Mikkel.

What do you mean by data format exactly? I may have changed some minor naming things, but I don't think it would be relevant for your purpose. If you give me an example you're using I can give you confirmation that it still works.

I definitely changed visual field / tracklet images formats. Can give you those if you like - feel free to try out the beta (be aware we now require miniforge / conda-forge as the only default channel):

conda create -n track -c trex-beta trex

**Reply by roaldarbol** [reply]
Simply what the column names are in exported data, and whether it's `csv` and/or `npy`, `h5`, etc. Basically what I'd need to know to ensure my [`read_trex`](https://roald-arboel.com/animovement/reference/read_trex.html) function (which brings the data into a standardised format, same for all tracking/pose estimation tools).

For now, it'll primarily be to import the positions, not the visual field/tracklet.

This is the sample data I currently use: https://github.com/roaldarbol/animovement/blob/main/tests/testthat/data/trex/beetle.csv.

**Reply by roaldarbol** [reply]
> Can give you those if you like - feel free to try out the beta (be aware we now require miniforge / conda-forge as the only default channel):

Would love to give it a spin soon! I struggled to figure out how to use the YOLO models which meant I didn't get any further - if you can provide some more details (or have written some docs in `dev`) that would be immensely helpful!

**Reply by roaldarbol** [reply]
Hi @mooch443, happy new year!
Would you be able to share examples of the data output from v2? Also, do they come with uncertainty estimates (confidence/likelihood/probability)? That would be really great since it enables users to do their own filtering processes based on the uncertainty estimates. If they do not, I would highly recommend that they be added (DLC, SLEAP, idtracker.ai, FreeMoCap now all come with these estimates which is great for interoperable data pipelines 😊).

**Reply by mooch443** [reply]
Hey there! I started writing a reply a couple times, hopefully this time I finish it :-D So I hadn't really thought much about exporting these before you wrote, but indeed it seems like I added those some time ago (at least I assume that this is what you're talking about):

`detection_p` => confidence for the detection
`detection_class` => the class index

And then you'd use, for pose, a choice of `poseX[n]` and `poseY[n]` fields to get the raw data from that. Likelihood is not something that I thought about - how are you / would you be using that?

All of these are per individual per frame, so depend on the tracking. There is currently not a way to get all the raw data out of it, though I am now thinking that maybe that'd be good too. Do you think that would be a huge benefit?

The data looks the same otherwise, just some additional fields (like the ones mentioned above).
[15locusts_locust0.csv](https://github.com/user-attachments/files/18642804/15locusts_locust0.csv)
The `poseX[n]` and `poseY[n]` are automatically added to the output if you have loaded a pose model.

Regarding tutorials or instructions, there's a preliminary version of that that I keep updating online. I'll publish the link here, but this link might not work for long or at all: https://trex.run/docs/docs2/tutorials.html

**Reply by roaldarbol** [reply]
Hey! 
`detection_p` sounds exactly like that! What is `detection_class`?  It seems that you are essentially now doing both centroid tracking, pose estimation and segmentation (outline), is that correctly understood?

Personally, I would probably split them out into separate output files. For the pose estimation I would recommend having a look at this discussion on [FreeMoCap](https://github.com/freemocap/freemocap/issues/653); and also for pose estimation, ensure that an estimate of uncertainty is exported for each keypoint for each frame. For just the poses (and the centroid and "front" too), it would be great to have a tidy export, see https://github.com/freemocap/freemocap/issues/653#issuecomment-2490548994.

I'm also curious how you are representing the segmentation data internally at the moment? Iø'm thinking about how to support analysis of those, how to best compress them etc., currently leaning towards some run length Encoding (RLE), but haven't continued down that path yet as I haven't encountered any tools/software that exports it - but if you'd consider it, we could try to think about it. :-)

Thanks for the tutorial link - I've saved it for my own future use in case it comes offline again. Nice illustrations by the way! :-D

**Reply by roaldarbol** [reply]
Regarding the segmentation data, I've started an issue for it over at animovement: https://github.com/roaldarbol/animovement/issues/108. Feel free to pitch in there!

**Reply by mooch443** [reply]
Hey!

> `detection_p` sounds exactly like that! What is `detection_class`?  It seems that you are essentially now doing both centroid tracking, pose estimation and segmentation (outline), is that correctly understood?

I am using pose data from YOLO to estimate posture (the trex outline midline combo) so trex can work with it just the same as the good old background subtraction. This means occasionally it will be necessary to provide 'pose_midline_indexes' to trex so it knows which pose points are the midline (plus head and tail in that order). For YOLO segmentation of course it's easy. Just using that in trex. For detection only the only thing you get is rectangles plus a class (detection_class) which is the index of the class assigned by YOLO. Like bicycle or table or nose. Detection p is then the probability of it assigned by YOLO.

> Personally, I would probably split them out into separate output files. For the pose estimation I would recommend having a look at this discussion on [FreeMoCap](https://github.com/freemocap/freemocap/issues/653); and also for pose estimation, ensure that an estimate of uncertainty is exported for each keypoint for each frame. For just the poses (and the centroid and "front" too), it would be great to have a tidy export, see https://github.com/freemocap/freemocap/issues/653#issuecomment-2490548994.

Thanks for the links, I will check! Right now I won't change output formats anymore, but I can add some still. Just not modify existing ones. For next version maybe :)

> I'm also curious how you are representing the segmentation data internally at the moment? Iø'm thinking about how to support analysis of those, how to best compress them etc., currently leaning towards some run length Encoding (RLE), but haven't continued down that path yet as I haven't encountered any tools/software that exports it - but if you'd consider it, we could try to think about it. :-) 

There really aren't many "good" representations of outlines it seems. There are bad ones. Mine is a bit compressed - so I save the absolute pixel coordinate of the first point, and then each outline point is a coordinate relative to the first point. This lets me store this in a smaller number container, although I am also scaling them dynamically so they fit even better! Outline coordinates are saved with precision depending on the overall spread of points. The worst case would be one outlier point and many many other points with little details closeby each other. But that basically can't happen in cohesive blobs.

> Thanks for the tutorial link - I've saved it for my own future use in case it comes offline again. Nice illustrations by the way!

Thanks! Although I have barely worked on those yet :D right now focusing on some workshops were giving. Take care!

**Reply by roaldarbol** [reply]
> This means occasionally it will be necessary to provide 'pose_midline_indexes' to trex so it knows which pose points are the midline (plus head and tail in that order)


Ah, so `pose_midline_index` is like what we could also call `individual`(in case they are animals) or `id`, etc. `detection_class` also makes complete sense to me now, thanks!

Let's chat more about potential extra output formats once the time comes - I think it's completely prudent choice to stick with what you've got. Maybe adding extras can be a way of testing out for a future version too. :-)

Regarding the segmentation representations, it sounds like you've given it a great deal of thought. Would you mind sharing some insights on [this issue](https://github.com/roaldarbol/animovement/issues/108)? (I know you're wrapping up the v2 release, so just afterwards). Maybe the group of us can come to some agreement, do some compression tests, etc.
